Move "MAC assumption" deeper

henriman opened the pull request:
At the moment, we simply assume that `hf_valid` holds after the comparison in `router/dataplane.go`, `verifyCurrentMAC`. This PR moves the assumption deeper into the code (into `path.FullMAC`), by instead assuming that the abstraction of the result of `path.FullMAC` equals the abstract computation of the MAC on the abstraction of the info and hop fields.

Besides slightly reducing the assumption, this change is necessary in order to be able to declassify the MAC tag when proving secure information flow (cf. #394, #409): This needs to happen before comparing it to the given tag (using `subtle.ConstantTimeCompare` in `verifyCurrentMAC`, `processOHP`) -- to satisfy low control flow, but also because the result of this comparison is exactly what is leaked about the router's key. Consequently, we need to abstractly describe that `FullMAC`/`MAC` compute such MAC tags s.t. we can allow declassification at this point.

---

### Approach

In order to keep track of which AS's secret key is used by the hash instances produced by `DataPlane.macFactory`/by the hash instance saved in `scionPacketProcessor.mac`, we define an extension of the `hash.Hash` interface called `ScionHashSpec` which allows us to access the corresponding AS identifier using function `ScionHashSpec.Asid`. 
- We decided to extend the interface instead of adding `Asid` directly to `hash.Hash` as this is specific to VerifiedSCION. Consequently (as e.g. `scionPacketProcessor.mac` needs to remain of type `hash.Hash`), we save the `ScionHashSpec` "version" of `mac` in `scionPacketProcessor.macSpec` and relate them using `mac === macSpec`.
- We establish that `Asid` returns the correct AS identifier in `SetKey` (using the appropriately adapted `InitMac` function). This information is then propagated for `p scionPacketProcessor` using `p.macSpec.Asid() == dp.Asid()`.

In `FullMAC`/`MAC`, besides `h hash.Hash` we then also pass `sh hash.ScionHashSpec` (where `h === sh`); this allows us to describe the corresponding abstract computation by passing `sh.Asid()` to `io.nextMsgtermSpec`.

See the comments below for further discussion explaining the need for this approach.

---

### Status

This is done. However, as I needed to adapt the contracts of `SetIA` and ´SetKey`, one _should_ adapt `router/dataplane_spec_test.gobra` to show that these contracts are satisfiable (although I am certain they are).

#### Adapting `router/dataplane_spec_test.gobra`. 

As the current tests don't call `SetIA` or `SetKey`, but rather set these fields directly, this requires a few modifications. The main issue that I came across was the fact that we pass full permissions to `(*DataPlane).Mem` (counting the permissions in the mutex) to these methods; consequently, after a call to e.g. `SetIA`, we are not able to establish that some properties dependent on unrelated fields of `DataPlane` still hold.

I have started adapting this test in branch [`mac-assumption-dataplane-test`](https://github.com/henriman/VerifiedSCION/tree/mac-assumption-dataplane-test) of my fork of VerifiedSCION (though I did not come far). The approach I have taken thus far was to change the contracts of `SetIA`, `SetKey` to only require partial permissions to `(*DataPlane).Mem` and full permissions to the fields that are actually modified.

However, this is not very clean. @jcp19 suggested the following approach instead:
- Introduce a spec/abstract version of `DataPlane` which contains only the relevant fields, without permissions (e.g. save a `seq` instead of a slice).
- Introduce a function `Abs` that returns such an abstraction for a given `DataPlane`.
- Then for each method that we want to annotate in this way, introduce a function that performs the corresponding modification on the abstraction. E.g. for `SetIA`, we could have the following function:
```go
ghost type DataPlaneSpec ghost struct {
    ...
    localIA addr.IA
    ...
}

ghost
pure func (d DataPlaneSpec) UpdateIA(ia addr.IA) DataPlaneSpec {
    return d[localIA = ia]
}
```
- We can then use these functions to describe what fields a method modifies and which remain the same in the following way:
```go
...
ensures e == nil ==> d.Abs() == old(d.Abs()).UpdateIA(ia)
...
func (d *DataPlane) SetIA(ia addr.IA) (e error)
```

In order to prove these contracts, it might be helpful to use the approach I initially used -- i.e. only passing full permissions to the fields that are actually modified -- in the contract of an `outline` block within the "top-level method".

henriman committed:
Add MWE for proving SIF

henriman committed:
Add MWE2

henriman committed:
Address PR feedback

henriman committed:
Fix package names and remove state.lastMsg3

jcp19 committed:
Merge branch 'master' into master

jcp19 committed:
Merge branch 'master' into master

henriman committed:
Lay out moving the "MAC assumption" into FullMAC

This relates the abstraction of the computed MAC tag to the one computed
from the abstracted arguments, s.t. we can assert `hf_valid` from there.

jcp19 committed:
Merge branch 'master' into master

jcp19 committed:
Merge branch 'master' into master

henriman committed:
Merge branch 'master' of https://github.com/henriman/VerifiedSCION into mac-assumption

henriman committed:
Merge branch 'master' of https://github.com/viperproject/VerifiedSCION into mac-assumption

henriman committed:
Remove MWEs here and implement work-around for package clauses

henriman committed:
Try reordering annotations

henriman committed:
Merge branch 'master' of https://github.com/viperproject/VerifiedSCION

henriman committed:
Finish verification and clean up

henriman committed:
Start including MAC

jcp19 commented on verification/dependencies/crypto/subtle/constant_time.gobra:
```diff
@@ -18,5 +18,9 @@ trusted
 requires sl.Bytes(y, 0, len(y))
 ensures  len(x) != len(y) ==> res == 0
+ensures res != 0 ==> 
+    unfolding sl.Bytes(x, 0, len(x)) in 
+    unfolding sl.Bytes(y, 0, len(y)) in 
+    forall i int :: { x[i] } { y[i] } 0 <= i && i < len(x) ==> x[i] == y[i] 
```
Please use the `GetByte` function defined in the package `verification/utils/slices/slices.gobra` to avoid the unfoldings here

jcp19 commented on router/dataplane.go:
```diff
@@ -2901,15 +2900,26 @@ func (p *scionPacketProcessor) verifyCurrentMAC( /*@ ghost dp io.DataPlaneSpec,
+
+	// Needed to "access" postcondition of `FullMAC`:
+	// @ unfold acc(sl.Bytes(fullMac, 0, len(fullMac)), R21)
+	// ==> absHF.HVF == io.nextMsgtermSpec(...)
```
please drop comments like this once you are done with the changes, as they make the code harder to read

jcp19 commented on pkg/slayers/path/io_msgterm_spec.gobra:
```diff
@@ -18,7 +18,7 @@ package path
 import (
 	"verification/io"
-	. "verification/utils/definitions"
+	. "verification/utils/definitions"	
```
```suggestion
	. "verification/utils/definitions"
```

jcp19 commented on pkg/slayers/path/mac.go:
```diff
@@ -69,6 +86,14 @@ func FullMAC(h hash.Hash, info InfoField, hf HopField, buffer []byte) (res []byt
 	//@ assert h.Size() >= 16
 	res = h.Sum(buffer[:0])[:16]
+
+	// This is our "MAC assumption", linking the abstraction of the concrete 
+	// MAC tag to the abstract computation of the MAC tag.
```
please move this comment to the line immediately before the call to `AssumeForIO`

jcp19 commented on pkg/slayers/path/mac.go:
```diff
@@ -49,9 +59,16 @@ func MAC(h hash.Hash, info InfoField, hf HopField, buffer []byte) [MacLen]byte {
+// @   	absMac == io.nextMsgtermSpec(as_, absHF.InIF2, absHF.EgIF2, absInf.AInfo.V, absInf.UInfo)
 // @ decreases
-func FullMAC(h hash.Hash, info InfoField, hf HopField, buffer []byte) (res []byte) {
+func FullMAC(h hash.Hash, info InfoField, hf HopField, buffer []byte /*@, ghost as_ io.IO_as @*/) (res []byte) {
```
do we need to pass the AS as a parameter? Can't we compute it from the infofield?

henriman replied:
It seems that the AS identifier is not actually contained in `InfoField` (neither in the concrete nor the abstract). 

The abstract *`HopField`* does provide the function `extract_asid` (for `IO_ahi`; `extr_asid` for `IO_HF`). However, I don't think it makes sense to take the AS id from there, as this would again introduce an unsensible assumption: 
- Consider an arbitrary number of `HopField`s `hf_i` which agree on `ExpTime`, `ConsIngress` and `ConsEgress`, but all have different `MAC`s, resulting in different `hf_i.ToIO_HF().extr_asid()`
- Then the postcondition `absMac == io.nextMsgtermSpec(hf_i.ToIO_HF().extr_asid(), ...)` would hold for all `hf_i`, which is clearly insensible (as `absMac` will be the same for all `hf_i`).

Another option would be to extract the AS id from `absMac`. This should theoretically contain the AS id corresponding to the key in `h hash.Hash`. However:
1. I think logically it should not be possible to get the AS id from a MAC tag (and I think the only reason why it is is due to the abstract "encoding" of the secret key, which simply wraps the AS id in an ADT).
2. (Without introducing an assumption everywhere we call `FullMAC` or `MAC`) the AS id we get from `absMac` is not related to `dp.Asid()`. But if we don't relate it to `dp.Asid()`, then we would need to allow declassifying all MAC tags `mac` s.t. `AbsMac(mac) == nextMsgtermSpec(getASID(AbsMac(mac)), ...)`, which allows declassifying MAC tags computed from other keys (other hash functions) as well.
	- I guess this could be OK, as the key is the only secret and thus MAC tags computed from other keys should be low already. Still, it doesn't really express what we want to

Intuitively, I think it would make the most sense to add a ghost method to the `hash.Hash` interface which returns the AS id corresponding to the secret key that the hash was initialized with. Then our postcondition would simply give us that the abstraction of the MAC tag (computed from `h hash.Hash`) equals the abstract MAC computation using the "same" key as `h` (as the AS id is used by `nextMsgtermSpec` to get the abstract key).
- This approach requires passing around the information that the AS id returned by the introduced ghost method (of the hash created in `SetKey`) equals `dp.Asid()`, which adds some complexity.
- Of course this would also be a rather SCION-specific annotation of `hash.Hash`

henriman replied:
We chose the latter option (i.e. introducing a method which returns an AS identifier). However, instead of adding this method to `hash.Hash`, we define the interface `hash.ScionHashSpec` which embeds `hash.Hash` and extends it with this method. This is now implemented.

henriman committed:
Address PR

I also adjusted the code to match the "new style conventions" (e.g. space between // and @)

henriman committed:
Start adding AS identifer to hash

henriman committed:
Save work

henriman committed:
Merge remote-tracking branch 'origin/master' into mac-assumption

henriman committed:
Apply name changes

jcp19 committed:
Merge branch 'master' into mac-assumption

henriman committed:
Merge branch 'master' of https://github.com/viperproject/VerifiedSCION into mac-assumption-scratch

henriman committed:
Make verify completely

henriman committed:
Save work

henriman committed:
Save work

henriman committed:
Merge branch 'master' of https://github.com/viperproject/VerifiedSCION into mac-assumption-scratch

henriman committed:
Finish introducing AsidMem

henriman committed:
General clean-up

henriman committed:
Minimize assertions

henriman committed:
Merge branch 'mac-assumption' into mac-assumption-scratch

henriman commented:
(`router/io-spec-lemmas.gobra` verifies successfully on my machine, so I'm not sure what the issue is with the pipeline.)

jcp19 commented:
@henriman are you using a different config for the router locally?

henriman committed:
Try to fix io-spec-lemmas.gobra

henriman committed:
Simplify fix

henriman commented:
> @henriman are you using a different config for the router locally?

It seems that the issue was due to me using a slightly outdated version of Gobra. With the most up-to-date version, I was able to reproduce and fix the issue (by adding an additional assert).

henriman committed:
Adapt dataplane_spec_test
